tools/generate-helpers/pg-table-bindings/store.go.tpl

```diff
@@ -378,13 +388,28 @@ func (s *storeImpl) UpsertMany(ctx context.Context, objs []*{{.Type}}) error {
     } else if !ok {
         return sac.ErrResourceAccessDenied
     }
-    {{- else if eq .ResourceType "globallyScoped" }}
+    {{- else if .Obj.IsGloballyScoped }}
     scopeChecker := sac.GlobalAccessScopeChecker(ctx).AccessMode(storage.Access_READ_WRITE_ACCESS).Resource(targetResource)
     if ok, err := scopeChecker.Allowed(ctx); err != nil {
         return err
     } else if !ok {
         return sac.ErrResourceAccessDenied
     }
+    {{- else if .Obj.IsDirectlyScoped -}}
+    scopeChecker := sac.GlobalAccessScopeChecker(ctx).AccessMode(storage.Access_READ_WRITE_ACCESS).Resource(targetResource)
+    if ok, err := scopeChecker.Allowed(ctx); err != nil {
+        return err
+    } else if !ok {
+        eas, err := scopeChecker.EffectiveAccessScope(permissions.Modify(targetResource))
+        if err != nil {
+            return err
+        }
+        for _, obj := range objs {
+            if !isInScope(obj, eas) {
+                return sac.ErrResourceAccessDenied
+            }
+        }
+    }
```

Comment:
If I understand correctly the logic here, we try to limit the number of checks first by looking if full access to the resource is granted. If not, it is assumed that access is partial and that finer-grain checks (item by item) can tell whether the input data is in scope or not.

Reply:
Correct